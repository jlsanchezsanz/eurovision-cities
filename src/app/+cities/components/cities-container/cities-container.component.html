<h1>Cities</h1>

<app-cities-table [cities]="cities$ | async"></app-cities-table>
