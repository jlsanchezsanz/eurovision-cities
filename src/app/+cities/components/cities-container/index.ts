export * from './cities-container.component';
