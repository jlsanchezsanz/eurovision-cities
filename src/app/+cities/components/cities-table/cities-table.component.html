<div class="mat-elevation-z8">
  <!-- Filter -->
  <section>
    <mat-form-field>
      <input
        matInput
        (keyup)="applyFilter($event.target.value)"
        placeholder="Filter"
      />
    </mat-form-field>
  </section>

  <!-- Table -->
  <section>
    <table mat-table [dataSource]="dataSource" matSort>
      <!-- Id Column -->
      <ng-container matColumnDef="id">
        <th mat-header-cell *matHeaderCellDef mat-sort-header>Id</th>
        <td mat-cell *matCellDef="let element">{{ element.id }}</td>
      </ng-container>

      <!-- Name Column -->
      <ng-container matColumnDef="name">
        <th mat-header-cell *matHeaderCellDef mat-sort-header>Name</th>
        <td mat-cell *matCellDef="let element">{{ element.name }}</td>
      </ng-container>

      <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
      <tr mat-row *matRowDef="let row; columns: displayedColumns"></tr>
    </table>
  </section>

  <!-- Pagination -->
  <section>
    <mat-paginator [pageSizeOptions]="pageSizeOptions" showFirstLastButtons>
    </mat-paginator>
  </section>
</div>
