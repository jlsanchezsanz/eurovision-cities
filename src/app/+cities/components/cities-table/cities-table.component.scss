.wrapper {
  padding: 10px;
}

table {
  width: 100%;
}
