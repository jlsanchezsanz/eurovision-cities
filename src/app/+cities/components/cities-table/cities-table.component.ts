import { Component, ChangeDetectionStrategy, Input } from '@angular/core';
import { City } from '../../../models';

@Component({
  selector: 'app-cities-table',
  templateUrl: './cities-table.component.html',
  styleUrls: ['./cities-table.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CitiesTableComponent {
  @Input() public cities: City[];

  public displayedColumns: string[] = ['id', 'name'];
}
