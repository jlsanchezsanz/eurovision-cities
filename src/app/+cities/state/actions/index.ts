export * from './cities.actions';
export * from './pagination.actions';
