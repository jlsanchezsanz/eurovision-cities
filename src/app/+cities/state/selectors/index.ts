export * from './cities.selectors';
export * from './pagination.selectors';
