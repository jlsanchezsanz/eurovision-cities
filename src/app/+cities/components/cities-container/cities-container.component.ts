import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { PageEvent } from '@angular/material';
import { Store, select } from '@ngrx/store';
import { Observable, combineLatest } from 'rxjs';

import { State } from '../../../state/reducers';
import { fetchCities, updatePage } from '../../state/actions';
import { City } from '../../../models';
import { getCities } from '../../state/selectors';
import {
  getPage,
  getPageSize,
  getTotalElements,
} from '../../state/selectors/pagination.selectors';
import { take } from 'rxjs/operators';

@Component({
  selector: 'app-cities-container',
  templateUrl: './cities-container.component.html',
  styleUrls: ['./cities-container.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CitiesContainerComponent implements OnInit {
  public cities$: Observable<City[]>;
  public page$: Observable<number>;
  public pageSizeOptions: number[] = [5, 10, 25, 100];
  public pageSize$: Observable<number>;
  public totalElements$: Observable<number>;

  constructor(private store: Store<State>) {}

  public ngOnInit(): void {
    this.cities$ = this.store.pipe(select(getCities));
    this.page$ = this.store.pipe(select(getPage));
    this.pageSize$ = this.store.pipe(select(getPageSize));
    this.totalElements$ = this.store.pipe(select(getTotalElements));
    this.fetchCities();
  }

  public onPageChange(pageEvent: PageEvent): void {
    const { pageIndex: page } = pageEvent;
    this.store.dispatch(updatePage({ page }));
  }

  private fetchCities(): void {
    combineLatest(this.page$, this.pageSize$)
      .pipe(take(1))
      .subscribe(([page, size]) =>
        this.store.dispatch(fetchCities({ page, size }))
      );
  }
}
