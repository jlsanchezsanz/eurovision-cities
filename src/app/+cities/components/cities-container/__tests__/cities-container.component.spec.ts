import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { MockComponent } from 'ng-mocks';

import { CitiesContainerComponent } from '../cities-container.component';
import { CitiesTableComponent } from '../../cities-table';
import { fetchCities } from '../../../state/actions';
import { getCities } from '../../../state/selectors';
import { citiesResponseMock } from '../../../../mocks';

describe('CitiesContainerComponent', () => {
  let component: CitiesContainerComponent;
  let fixture: ComponentFixture<CitiesContainerComponent>;
  let store: MockStore;
  const initialState = {};

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        CitiesContainerComponent,
        MockComponent(CitiesTableComponent),
      ],
      providers: [
        provideMockStore({
          initialState,
          selectors: [
            { selector: getCities, value: citiesResponseMock.content },
          ],
        }),
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CitiesContainerComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should fetch cities', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch');
    fixture.detectChanges();
    expect(dispatchSpy).toBeCalledTimes(1);
    expect(dispatchSpy).toBeCalledWith(fetchCities());
  });
});
