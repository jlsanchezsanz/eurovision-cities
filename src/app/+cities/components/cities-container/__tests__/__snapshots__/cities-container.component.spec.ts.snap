// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesContainerComponent should match snapshot 1`] = `
<app-cities-container
  cities$={[Function Store]}
  page$={[Function Store]}
  pageSize$={[Function Store]}
  pageSizeOptions={[Function Array]}
  store={[Function MockStore]}
  totalElements$={[Function Store]}
>
  <header>
    <h1>
      Cities
    </h1>
  </header><section>
    <app-cities-table />
  </section><section>
    <app-cities-paginator />
  </section>
</app-cities-container>
`;
